backend/GameManager.py

```diff
@@ -58,5 +82,173 @@ def get_pieces_left(self):
+        if not self.selected_piece:
+            Exception("MoveError -- No piece selected")
+
+        start_row, start_col = self.selected_piece
+
+        # Validate the target position is empty and adjacent
+        if self.board.check_position(target_row, target_col) != Color.EMPTY:
+            Exception("MoveError -- Target position is not empty")
+
+        if not self.can_fly(self.turn):
+            if not self.is_adjacent_and_empty(start_row, start_col, target_row, target_col):
+                Exception("MoveError -- Invalid move, pieces can only move to adjacent positions")
+
+        # Perform the move
+        self.board.set_position(target_row, target_col, self.turn.name.lower())
+        self.board.set_position(start_row, start_col, Color.EMPTY)
+        # self.board.board[start_row][start_col] = None
+
+        # removing the new position from open move and adding the previous position to it
+        self.open_moves.append((start_row, start_col))
+        self.open_moves.remove((target_row, target_col))
+
+        self.selected_piece = None
+        return True
+
+    def get_piece_at(self, row, col):
+        return self.board.board[row][col]  
+    
+    # def is_mill_formed(self, row, col, turn):
+    #     for mill in self.mills:
+    #         if (row, col) in mill:
+    #             # Check if all three positions in this mill belong to the current player
+    #             first = self.get_piece_at(r, c)
+                
+    #             if all(self.get_piece_at(r, c) == turn for r, c in mill):
+    #                 return True
+    #     return False
+
+
+    def is_mill_formed(self, row, col):
+        for mill in self.mills:
+            if (row, col) in mill:
+                # Create an empty list to store comparison results
+                pieces_in_mill = []
+                
+                # Loop through each position in the mill
+                for r, c in mill:
+                    piece_at_position = self.get_piece_at(r, c)  # Get the piece at the mill position
+                    is_current_turn_piece = (piece_at_position.get_state().name == self.get_turn().name)  # Check if it's the current player's turn
+                    pieces_in_mill.append(is_current_turn_piece)  # Append the result (True/False) to the list
+                
+                # Now use 'all' to check if all the values in the list are True
+                if all(pieces_in_mill):
+                    return True
+
+        return False
+    
+    def remove_piece_mill(self, row, col):
+        self.open_moves.append((row, col))
+        self.board.set_position(row, col, Color.EMPTY)
+        self.waiting_for_removal = False  # Reset the removal state
+        self.removable_pieces = []  # Clear the list of removable pieces
+        self.end_turn()  # End the turn after removal
+        self.end_turn()  # swap the turn to opponent
+        return True
+    
+    def remove_opponent_piece(self , pieces_on_board):
```

Comment:
Actually, why is `pieces_on_board` a local variable?? Shouldn't we be storing that as part of the instance?

Reply:
It'd be much easier to test if this weren't stored in the GUI. 😅

Reply:
I would expect this to only have a row and col as input to remove opponent piece for this function. @suryansh-max .